c_emulator/riscv_rvfi_callbacks.c

```diff
@@ -0,0 +1,38 @@
+#include "riscv_config.h"
+
+int zrvfi_write(uint64_t addr, int64_t width, lbits value);
+int zrvfi_read(uint64_t addr, sail_int width, lbits value);
+int zrvfi_mem_exception(uint64_t addr);
+int zrvfi_wX(int64_t reg, uint64_t value);
+
+int mem_write_callback(uint64_t addr, uint64_t width, lbits value)
+{
+  if (rv_enable_callbacks)
+    zrvfi_write(addr, width, value);
+}
+int mem_read_callback(const char *type, uint64_t addr, uint64_t width,
+                      lbits value)
+{
+  if (rv_enable_callbacks) {
```

Comment:
We want a single callback that does *both* the printing, and the RVFI stuff (if enabled). So the contents of this file should be merged into `riscv_default_callbacks.c`. Something like this:

```
int mem_write_callback(uint64_t addr, uint64_t width, lbits value)
{
  if (config_print_mem_access) {
    char *lbits_data = get_lbits_data(value);
    printf("mem[0x%.16lX] <- 0x", addr);
    for (int i = width - 1; i >= 0; --i)
      printf("%02hhX", lbits_data[i]);
    printf("\n");
    free(lbits_data);
  }

#ifdef RVFI_DII
  if (config_enable_rvfi) {
    sail_int len;
    CREATE(sail_int)(&len);
    CONVERT_OF(sail_int, mach_int)(&len, width);
    zrvfi_read(addr, len, value);
    KILL(sail_int)(&len);
  }
#endif
}
```

In future we should remove the `#ifdef` so RVFI_DII support is always enabled. It's unnecessary to have it as a compile time option - it just doubles compilation time. (But don't do it in this PR.)